File: backend/utils/rate_limit.py
"""In-memory sliding-window rate limiter for OTP send endpoint."""
import time
from collections import defaultdict, deque
from threading import Lock

_STORE: dict[str, deque] = defaultdict(deque)
_LOCK = Lock()


def check_and_record(key: str, limit: int, window_seconds: int) -> tuple[bool, int]:
    """Return (allowed, remaining). If allowed=False, remaining is seconds until reset."""
    now = time.time()
    with _LOCK:
        dq = _STORE[key]
        cutoff = now - window_seconds
        while dq and dq[0] < cutoff:
            dq.popleft()
        if len(dq) >= limit:
            reset_in = int(window_seconds - (now - dq[0])) + 1
            return False, reset_in
        dq.append(now)
        return True, limit - len(dq)

File: backend/src/utils/rateLimit.js
const store = new Map();

const checkAndRecord = (key, limit, windowSeconds) => {
  const now = Date.now() / 1000;
  if (!store.has(key)) {
    store.set(key, []);
  }

  const times = store.get(key);
  const cutoff = now - windowSeconds;

  // Remove timestamps outside window
  while (times.length > 0 && times[0] < cutoff) {
    times.shift();
  }

  if (times.length >= limit) {
    const resetIn = Math.max(1, Math.ceil(windowSeconds - (now - times[0])));
    return { allowed: false, remaining: resetIn };
  }

  times.push(now);
  return { allowed: true, remaining: limit - times.length };
};

module.exports = {
  checkAndRecord,
};
